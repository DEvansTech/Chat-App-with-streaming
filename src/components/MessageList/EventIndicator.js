import React from 'react';
import styled from '@stream-io/styled-components';
import PropTypes from 'prop-types';

import { withTranslationContext } from '../../context';
import { Avatar } from '../Avatar';

const Date = styled.Text`
  font-size: 10;
  color: rgba(0, 0, 0, 0.5);
  ${({ theme }) => theme.messageList.eventIndicator.date.css}
`;

const MemberUpdateContainer = styled.View`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  padding: 10px;
  ${({ theme }) => theme.messageList.eventIndicator.memberUpdateContainer.css}
`;

const MemberUpdateTextContainer = styled.View`
  display: flex;
  flex-direction: column;
  margin-left: 10px;
  ${({ theme }) =>
    theme.messageList.eventIndicator.memberUpdateTextContainer.css}
`;

const MemberUpdateText = styled.Text`
  font-size: 13px;
  font-style: italic;
  color: rgba(0, 0, 0, 0.5);
  ${({ theme }) => theme.messageList.eventIndicator.memberUpdateText.css}
`;

/**
 * A component to display a message regarding channel notifications such as
 * 'member.added', 'member.removed' etc.
 */
const EventIndicator = ({ event, t, tDateTimeParser }) => {
  if (event.type === 'member.added' || event.type === 'member.removed') {
    return (
      <MemberUpdateContainer>
        <Avatar name={event.user.name} image={event.user.image} />
        <MemberUpdateTextContainer>
          <MemberUpdateText>
            {event.type === 'member.added'
              ? t('{{ username }} joined the chat', {
                  username: event.user.name,
                })
              : t('{{ username }} was removed from the chat', {
                  username: event.user.name,
                })}
          </MemberUpdateText>
          <Date>{tDateTimeParser(event.received_at).format('LT')}</Date>
        </MemberUpdateTextContainer>
      </MemberUpdateContainer>
    );
  }

  return null;
};

EventIndicator.propTypes = {
  event: PropTypes.object,
};

export default withTranslationContext(EventIndicator);
