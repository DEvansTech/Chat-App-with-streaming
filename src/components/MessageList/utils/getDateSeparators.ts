import type { PaginatedMessageListContextValue } from '../../../contexts/paginatedMessageListContext/PaginatedMessageListContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export type GetDateSeparatorsParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  messages:
    | PaginatedMessageListContextValue<At, Ch, Co, Ev, Me, Re, Us>['messages']
    | ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>['threadMessages'];
  hideDateSeparators?: boolean;
};

export type DateSeparators = {
  [key: string]: Date;
};

export const getDateSeparators = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  params: GetDateSeparatorsParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { hideDateSeparators, messages } = params;
  const dateSeparators: DateSeparators = {};

  if (hideDateSeparators) {
    return dateSeparators;
  }

  for (let i = 0; i < messages.length; i++) {
    const previousMessage = messages[i - 1];
    const message = messages[i];

    if (
      message.type === 'channel.event' ||
      message.type === 'message.read' ||
      message.deleted_at
    ) {
      continue;
    }

    const messageDate = message.created_at.getDay();

    const prevMessageDate = previousMessage
      ? previousMessage.created_at.getDay()
      : messageDate;

    if (i === 0 || messageDate !== prevMessageDate) {
      dateSeparators[message.id] = message.created_at;
    }
  }

  return dateSeparators;
};
