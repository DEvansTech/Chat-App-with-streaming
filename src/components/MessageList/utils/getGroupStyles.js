export const getGroupStyles = (messagesWithDates, noGroupByUser) => {
  const numberOfMessages = messagesWithDates.length;
  const messageGroupStyles = {};

  const messages = [...messagesWithDates];

  for (let i = 0; i < numberOfMessages; i++) {
    const previousMessage = messages[i - 1];
    const message = messages[i];
    const nextMessage = messages[i + 1];
    const groupStyles = [];

    if (message.type === 'channel.event') {
      continue;
    }

    if (message.type === 'message.date') {
      continue;
    }

    const userId = message.user ? message.user.id : null;

    const isTopMessage =
      !previousMessage ||
      previousMessage.type === 'message.date' ||
      previousMessage.type === 'system' ||
      previousMessage.type === 'channel.event' ||
      (previousMessage.attachments &&
        previousMessage.attachments.length !== 0) ||
      userId !== previousMessage.user.id ||
      previousMessage.type === 'error' ||
      previousMessage.deleted_at;

    const isBottomMessage =
      !nextMessage ||
      nextMessage.type === 'message.date' ||
      nextMessage.type === 'system' ||
      nextMessage.type === 'channel.event' ||
      (nextMessage.attachments && nextMessage.attachments.length !== 0) ||
      userId !== nextMessage.user.id ||
      nextMessage.type === 'error' ||
      nextMessage.deleted_at;

    if (isTopMessage) {
      groupStyles.push('top');
    }

    if (isBottomMessage) {
      if (isTopMessage || message.deleted_at || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.push('bottom');
      }
    }

    if (!isTopMessage && !isBottomMessage) {
      if (message.deleted_at || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('middle');
      }
    }

    if (message.attachments.length !== 0) {
      groupStyles.splice(0, groupStyles.length);
      groupStyles.push('single');
    }

    if (noGroupByUser) {
      groupStyles.splice(0, groupStyles.length);
      groupStyles.push('single');
    }

    messageGroupStyles[message.id] = groupStyles;
  }

  return messageGroupStyles;
};
