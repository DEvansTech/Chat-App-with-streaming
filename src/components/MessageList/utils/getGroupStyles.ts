import type { DateSeparators } from './getDateSeparators';

import type { GroupType } from '../hooks/useMessageList';

import type { PaginatedMessageListContextValue } from '../../../contexts/paginatedMessageListContext/PaginatedMessageListContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';
import type { FormatMessageResponse } from 'stream-chat';

export type GetGroupStylesParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  dateSeparators: DateSeparators;
  messages:
    | PaginatedMessageListContextValue<At, Ch, Co, Ev, Me, Re, Us>['messages']
    | ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>['threadMessages'];
  hideDateSeparators?: boolean;
  maxTimeBetweenGroupedMessages?: number;
  noGroupByUser?: boolean;
  userId?: string;
};

export const getGroupStyles = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  params: GetGroupStylesParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    dateSeparators,
    hideDateSeparators,
    maxTimeBetweenGroupedMessages,
    messages,
    noGroupByUser,
    userId,
  } = params;
  const messageGroupStyles: { [key: string]: GroupType[] } = {};

  const messagesFilteredForNonUser = messages.filter(
    (message) => !message.deleted_at || userId === message.user?.id,
  );

  for (let i = 0; i < messagesFilteredForNonUser.length; i++) {
    const previousMessage = messagesFilteredForNonUser[i - 1] as
      | typeof messagesFilteredForNonUser[0]
      | undefined;
    const message = messagesFilteredForNonUser[i];
    const nextMessage = messagesFilteredForNonUser[i + 1] as
      | typeof messagesFilteredForNonUser[0]
      | undefined;
    const groupStyles: GroupType[] = [];

    const userId = message?.user?.id || null;

    const isTopMessage =
      !previousMessage ||
      previousMessage.type === 'system' ||
      userId !== previousMessage?.user?.id ||
      previousMessage.type === 'error' ||
      !!previousMessage.deleted_at ||
      (!hideDateSeparators && dateSeparators[message.id]) ||
      messageGroupStyles[previousMessage.id]?.includes('bottom');

    const isBottomMessage =
      !nextMessage ||
      nextMessage.type === 'system' ||
      userId !== nextMessage?.user?.id ||
      nextMessage.type === 'error' ||
      !!nextMessage.deleted_at ||
      (!hideDateSeparators && dateSeparators[nextMessage.id]) ||
      (maxTimeBetweenGroupedMessages !== undefined &&
        nextMessage.created_at.getTime() - message.created_at.getTime() >
          maxTimeBetweenGroupedMessages);

    /**
     * Add group styles key for top message
     */
    if (isTopMessage) {
      groupStyles.push('top');
    }

    /**
     * Add group styles key for bottom message
     */
    if (isBottomMessage) {
      /**
       * If the bottom message is also the top, or deleted, or an error,
       * add the key for single message instead of bottom
       */
      if (isTopMessage || message.deleted_at || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.push('bottom');
      }
    }

    /**
     * Add the key for all non top or bottom messages, if the message is
     * deleted or an error add the key for single otherwise middle
     */
    if (!isTopMessage && !isBottomMessage) {
      if (message.deleted_at || message.type === 'error') {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('single');
      } else {
        groupStyles.splice(0, groupStyles.length);
        groupStyles.push('middle');
      }
    }

    /**
     * If noGroupByUser set add the key for single
     */
    if (noGroupByUser) {
      groupStyles.splice(0, groupStyles.length);
      groupStyles.push('single');
    }

    if (message.id) {
      messageGroupStyles[message.id] = groupStyles;
    }
  }

  return messageGroupStyles;
};
