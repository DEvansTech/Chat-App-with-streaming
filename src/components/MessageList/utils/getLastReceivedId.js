export const getLastReceivedId = (messages) => {
  const l = messages.length;
  let lastReceivedId = null;

  for (let i = l; i > 0; i--) {
    if (
      messages[i] !== undefined &&
      messages[i].status !== undefined &&
      messages[i].status === 'received'
    ) {
      lastReceivedId = messages[i].id;
      break;
    }
  }

  return lastReceivedId;
};
