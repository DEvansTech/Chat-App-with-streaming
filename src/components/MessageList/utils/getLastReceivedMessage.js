export const getLastReceivedMessage = (messages) => {
  let lastReceivedMessage = null;

  /**
   * There are no status on dates so they will be skipped
   */
  for (let i = 0; i < messages.length; i++) {
    if (
      messages[i] !== undefined &&
      messages[i].id !== undefined &&
      messages[i].status !== undefined &&
      messages[i].status === 'received'
    ) {
      lastReceivedMessage = messages[i];
      break;
    }
  }

  return lastReceivedMessage;
};
