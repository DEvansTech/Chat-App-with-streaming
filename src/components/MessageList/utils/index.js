export { getGroupStyles } from './getGroupStyles';
export { getLastReceivedMessage } from './getLastReceivedMessage';
export { getReadStates } from './getReadStates';
export { insertDates } from './insertDates';
