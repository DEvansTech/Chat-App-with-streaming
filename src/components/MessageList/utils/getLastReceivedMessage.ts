import { InsertDatesResponse, isDateSeparator } from '../utils/insertDates';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export const getLastReceivedMessage = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  messages: InsertDatesResponse<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  let lastReceivedMessage = null;

  /**
   * There are no status on dates so they will be skipped
   */
  for (let i = 0; i < messages.length; i++) {
    const message = messages[i];
    if (
      message !== undefined &&
      !isDateSeparator<At, Ch, Co, Ev, Me, Re, Us>(message) &&
      message.status !== undefined &&
      message.status === 'received'
    ) {
      lastReceivedMessage = messages[i];
      break;
    }
  }

  return lastReceivedMessage;
};
