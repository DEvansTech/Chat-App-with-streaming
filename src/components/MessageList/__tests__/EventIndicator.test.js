import React from 'react';
import { ThemeProvider } from '@stream-io/styled-components';
import { cleanup, render, waitFor } from '@testing-library/react-native';
import { generateStaticUser } from 'mock-builders';

import { TranslationContext } from '../../../context';
import EventIndicator from '../EventIndicator';
import { defaultTheme } from '../../../styles/theme';
import { Streami18n } from '../../../utils';

afterEach(cleanup);

describe('EventIndicator', () => {
  it('should render event indicator', async () => {
    const t = jest.fn((key) => key);
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const event = {
      received_at: new Date().toString(),
      type: 'member.added',
      user: generateStaticUser(0),
    };
    const { queryByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <TranslationContext.Provider value={{ ...translators, t }}>
          <EventIndicator event={event} />
        </TranslationContext.Provider>
      </ThemeProvider>,
    );

    expect(t).toHaveBeenCalledWith('{{ username }} joined the chat', {
      username: event.user.name,
    });
    await waitFor(() => {
      expect(queryByTestId('event-indicator')).toBeTruthy();
    });
  });

  it('should not render', async () => {
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const event = {
      received_at: new Date().toString(),
      type: 'member.whatever',
      user: generateStaticUser(0),
    };
    const { queryByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <TranslationContext.Provider value={translators}>
          <EventIndicator event={event} />
        </TranslationContext.Provider>
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(queryByTestId('event-indicator')).toBeFalsy();
    });
  });

  it('should match the snapshot for member removed', async () => {
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const event = {
      received_at: '2020-05-05T14:50:00',
      type: 'member.removed',
      user: generateStaticUser(1),
    };
    const { toJSON } = render(
      <ThemeProvider theme={defaultTheme}>
        <TranslationContext.Provider value={translators}>
          <EventIndicator event={event} />
        </TranslationContext.Provider>
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
