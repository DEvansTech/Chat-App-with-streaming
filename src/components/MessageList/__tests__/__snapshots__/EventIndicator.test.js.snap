// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventIndicator should match the snapshot for member removed 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
    ]
  }
  testID="event-indicator"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
      ]
    }
  >
    <Image
      accessibilityLabel="initials"
      onError={[Function]}
      resizeMethod="resize"
      size={32}
      source={
        Object {
          "uri": "https://i.imgur.com/SLx06PP.png",
        }
      }
      style={
        Array [
          Object {
            "borderRadius": 16,
            "height": 32,
            "width": 32,
          },
        ]
      }
      testID="avatar-image"
    />
  </View>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "column",
          "marginLeft": 10,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.5)",
            "fontSize": 13,
            "fontStyle": "italic",
          },
        ]
      }
    >
      Grant was removed from the chat
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.5)",
            "fontSize": 10,
          },
        ]
      }
    >
      2:50 PM
    </Text>
  </View>
</View>
`;
